packages/angular/ssr/src/app-engine.ts

```diff
@@ -0,0 +1,151 @@
+   * the cache is bypassed and a new instance is created for each request.
+   *
+   * If the request URL appears to be for a file (excluding `/index.html`), the method returns `null`.
+   * A request to `https://www.example.com/page/index.html` will render the Angular route
+   * corresponding to `https://www.example.com/page`.
+   *
+   * @param {Request} request - The incoming HTTP request object to be rendered.
+   * @param {unknown} [requestContext] - Optional additional context for the request, such as metadata.
+   * @returns {Promise<Response | null>} A promise that resolves to a Response object or null if the request is file-like.
+   */
+  async render(request: Request, requestContext?: unknown): Promise<Response | null> {
+    // Skip if the request looks like a file but not `/index.html`.
+    const { pathname } = new URL(request.url);
+    if (isFileLike(pathname) && !pathname.endsWith('/index.html')) {
+      return null;
+    }
+
+    const [locale, loadModule] = this.getEntryPointFromUrl(request.url);
+    let serverApp = this.appsCache.get(locale);
+    if (!serverApp || AngularAppEngine.isDevMode) {
+      const { AngularServerApp } = await loadModule();
+
+      serverApp = new AngularServerApp({
+        isDevMode: AngularAppEngine.isDevMode,
+        hooks: this.hooks,
+      });
+
+      this.appsCache.set(locale, serverApp);
+    }
+
+    return serverApp.render(request, requestContext);
+  }
+
+  /**
+   * Retrieves the Angular server application's entry point path and locale based on the requested URL.
+   *
+   * @param {string} url - The URL from which to derive the locale and determine the entry point.
+   * @returns {[string, () => Promise<{ AngularServerApp: typeof AngularServerApp }>]} An array where:
+   * - The first element is the locale derived from the URL.
+   * - The second element is a function that returns a promise resolving to an object containing the `AngularServerApp` type.
+   * @throws {Error} If the entry point path cannot be determined for the given URL.
+   */
+  private getEntryPointFromUrl(url: string): [
+    locale: string,
+    loadModule: () => Promise<{
+      AngularServerApp: typeof AngularServerApp;
+    }>,
+  ] {
+    // Find bundle for locale
+    const entryPoints = this.entryPoints;
+    if (entryPoints.size === 1) {
+      return entryPoints.entries().next().value;
+    }
+
+    const potentialLocale = getPotentialLocaleIdFromUrl(url, this.manifest.basePath);
+    const entryPoint = entryPoints.get(potentialLocale);
+    if (!entryPoint) {
+      throw new Error(`Unable to determine Angular application entrypoint for: ${url}.`);
+    }
+
+    return [potentialLocale, entryPoint];
+  }
+}
+
+/**
+ * Determines if the given pathname corresponds to a file-like resource.
+ *
+ * @param {string} pathname - The pathname to check.
+ * @returns {boolean} True if the pathname appears to be a file, false otherwise.
+ */
+function isFileLike(pathname: string): boolean {
```

Comment:
In this case, we want to avoid rendering non-existent files using Angular. If we do, all non-existent files will result in returning a server-side rendered (SSR) version of the home page instead of a 404 error.

Generally, having an extension, especially a non-HTML one, for a Single Page Application (SPA) is considered bad practice. I have provided a brief summary on this issue in [this GitHub comment](https://github.com/angular/angular-cli/issues/27905#issuecomment-2185785230), which is also the current behavior of the dev server.

Reply:
I was also curious whether we may have enough info in the manifest to detect whether it's a known route vs a request for an unknown resource.

Reply:
Or maybe we remove it now, and integrate the route matcher earlier.

Reply:
Up to you exactly when is the right time to handle this. I do think it would be good to not rely on the file extension for this use case eventually, if that's possible.

Reply:
I do prefer to handle this in a separate PR that I aim to start working on once this is landed. Primarily because this PR is becoming rather large and would make it harder to review.